### PR TITLE
Add debug docs

### PR DESCRIPTION
**Pull request by mrocklin**
cc @martindurant @stuartarchibald

### CODE REVIEW

**Comment by mrocklin** [comment]
Feedback appreciated

File: docs/source/debugging.rst [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+However, you can not use the ``pdb`` module or ``%debug`` IPython magics with
+these tracebacks to look at the value of variables during failure.  You can
+only inspect things visually.  Additionally, the top of the traceback may be
+filled with functions that are dask-specific and not relevant to your
+problem, you can safely ignore these.
+
+Both the single-machine and distributed schedulers do this.
+
+
+Use the Single-Threaded Scheduler
```
Comment (martindurant) [inline docs/source/debugging.rst]: A caveat against blindly accessing large datasets that might not fit in local memory?

File: docs/source/debugging.rst [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+.. code-block:: python
+
+   x.compute(rerun_exceptions_locally=True)
+
+On the distributed scheduler use the ``recreate_error_locally`` method on
+anything that contains ``Futures`` :
+
+.. code-block:: python
+
+   >>> x.compute()
```
Comment (martindurant) [inline docs/source/debugging.rst]: Here x may be pure delayed and not contain futures, unless it is persisted.. Better might be `f = client.compute(x)`, `client.recreate_error_locally(f)`.

File: docs/source/debugging.rst [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+Web Diagnostics
+~~~~~~~~~~~~~~~
+
+First, the distributed scheduler has a number of diagnostic web pages showing
+dozens of recorded metrics like CPU, memory, network, and disk use, a history
+of previous tasks, allocation of tasks to workers, worker memory pressure, work
+stealing, open file handle limits, etc..  *Many* problems can be correctly
+diagnosed by inspecting these pages.  By default these are available at
+``http://scheduler:8787/`` ``http://scheduler:8788`` and ``http://worker:8789``
```
Comment (martindurant) [inline docs/source/debugging.rst]: Link to UI pages, or bries description of what can be found in each location?

File: docs/source/debugging.rst [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+using `LocalCluster <http://distributed.readthedocs.io/en/latest/local-cluster.html>`_
+which is what runs when you just call ``Client()``
+
+.. code-block:: python
+
+   >>> from dask.distributed import Client, LocalCluster
+   >>> client = Client()  # This is actually the following two commands
+
+   >>> cluster = LocalCluster()
+   >>> client = Client(cluster.scheduler.address)
```
Comment (martindurant) [inline docs/source/debugging.rst]: The client has no reference to the cluster, right? So you need the two lines, not the one, to get access to the scheduler state.

Comment (mrocklin) [inline docs/source/debugging.rst]: The client does have a reference at `client.cluster`, but when debugging I find it's nice to have the cluster constructor exposed.  That way you can control keywords on it.  In these docs it also makes it more explicit that there is another object in play.

**Comment by martindurant** [comment]
Is it worth saying something about map/submit of futures, and just checking their status? That's the simplest way of at least avoiding errored data.

File: docs/source/debugging.rst [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+On the distributed scheduler use the ``recreate_error_locally`` method on
+anything that contains ``Futures`` :
+
+.. code-block:: python
+
+   >>> x.compute()
+   ZeroDivisionError(...)
+
+   >>> %pdb
+   >>> client.recreate_error_locally(x)
```
Comment (mrocklin) [inline docs/source/debugging.rst]: @martindurant any thoughts on how to bring together these APIs somehow?  Perhaps even just the names?

Comment (martindurant) [inline docs/source/debugging.rst]: You mean recreate_error_locally->rerun_exceptions_locally (or vice versa) ? They do slightly different things, not sure we want to confuse people.

Comment (mrocklin) [inline docs/source/debugging.rst]: Yes, something like that.  The intent seems the same to me even if the internals are necessarily different.  Is there particular confusion that you expect?

We probably can't change the dask/dask name at this stage (it has been around for a while) but could change the dask/distributed name.

File: docs/source/debugging.rst (hunk 3 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: Added a link slightly higher here.

File: docs/source/debugging.rst (hunk 2 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: done

File: docs/source/debugging.rst (hunk 1 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: done

**Comment by mrocklin** [comment]
> Is it worth saying something about map/submit of futures, and just checking their status? That's the simplest way of at least avoiding errored data.

Done using `futures_of`.  Generally we don't need to use map/submit to get futures, that's just the API in which they are produced in a bare/raw form.  All client methods deal in futures, even those that produce collections.

**Comment by mrocklin** [comment]
Any additional comments @martindurant or @stuartarchibald ?

File: docs/source/debugging.rst [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By default, Dask already copies the exception and traceback wherever they
+occur and reraises that exception locally.  If your task failed with a
+``ZeroDivisionError`` remotely then you'll get a ``ZeroDivisionError`` in your
+interactive session.  Similarly you'll see a full traceback of where this error
+occurred, which, just like in normal Python, can help you to identify the
+troublsome spot in your code.
+
+However, you can not use the ``pdb`` module or ``%debug`` IPython magics with
```
Comment (martindurant) [inline docs/source/debugging.rst]: cannot

**Comment by mrocklin** [comment]
In private conversation @stuartarchibald mentioned that he's busy until Tuesday but would be happy to comment then.  I'm leaving this up here until then.

File: docs/source/debugging.rst [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+=========
+
+Debugging parallel programs is hard.  Normal debugging tools like logging and
+using ``pdb`` to interact with tracebacks stop working normally when exceptions
+occur in far-away machines or different processes or threads.
+
+Dask has a variety of mechanisms to make this process easier.  Depending on
+your situation some of these approaches may be more appropriate than others.
+
+These approaches are ordered from light weight or easy solutions to more
```
Comment (stuartarchibald) [inline docs/source/debugging.rst]: `s/light weight/lightweight/`

File: docs/source/debugging.rst [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+
+Normally when a computation computes in a separate thread or a different
+machine these approaches break down.  Dask provides a few mechanisms to
+recreate the normal Python debugging experience.
+
+Inspect Exceptions and Tracebacks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By default, Dask already copies the exception and traceback wherever they
+occur and reraises that exception locally.  If your task failed with a
```
Comment (stuartarchibald) [inline docs/source/debugging.rst]: I think this assumes that the exception is serializable, if it doesn't you get an infinite recursion error from the depths of pickle.

File: docs/source/debugging.rst [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+Dask ships with a simple single-threaded scheduler.  This doesn't offer any
+parallel performance improvements, but does run your Dask computation
+faithfully in your local thread, allowing you to use normal tools like ``pdb``,
+``%debug`` IPython magics, the profiling tools like the ``cProfile`` module and
+`snakeviz <https://jiffyclub.github.io/snakeviz/>`_.  This allows you to use
+all of your normal Python debugging tricks in Dask computations, as long as you
+don't need parallelism.
+
+This only works for single-machine schedulers.  It does not work with
+dask.distributed unless you use comfortable using the Tornado API (want to look
```
Comment (stuartarchibald) [inline docs/source/debugging.rst]: `s/you use/you are/`

Comment (stuartarchibald) [inline docs/source/debugging.rst]: `s/want to//` ?

File: docs/source/debugging.rst [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+at the `testing infrastructure
+<http://distributed.readthedocs.io/en/latest/develop.html#writing-tests>`_
+docs, which accomplish this).  Also, because this operates on a single machine
+it assumes that your computation can run on a single machine without exceeding
+RAM limits.  It may be wise to use this approach on smaller versions of your
```
Comment (stuartarchibald) [inline docs/source/debugging.rst]: `s/RAM/memory/` cgroups or something might restrict memory but RAM is still free?

File: docs/source/debugging.rst [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+This is a bit of a hack, but often practical when first exploring messy data.
+If you are using the concurrent.futures API (map, submit, gather) then this
+approach is more natural.
+
+
+Inspect Scheduling State
+------------------------
+
+Not all errors present themselves as Exceptions.  For example in a distributed
+systems workers may die unexpectedly or your computation may be unreasonably
```
Comment (stuartarchibald) [inline docs/source/debugging.rst]: Is it worth mentioning adding `with log_errors():` context management to `distributed/worker.py::apply_function`? Here might not be the most appropriate place though?

File: docs/source/debugging.rst [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+Web Diagnostics
+~~~~~~~~~~~~~~~
+
+First, the distributed scheduler has a number of `diagnostic web pages
+<http://distributed.readthedocs.io/en/latest/web.html>`_ showing dozens of
+recorded metrics like CPU, memory, network, and disk use, a history of previous
+tasks, allocation of tasks to workers, worker memory pressure, work stealing,
+open file handle limits, etc..  *Many* problems can be correctly diagnosed by
+inspecting these pages.  By default these are available at
+``http://scheduler:8787/`` ``http://scheduler:8788`` and ``http://worker:8789``
```
Comment (stuartarchibald) [inline docs/source/debugging.rst]: Worth saying what's at these default ports ? Also the 8787 has a trailing slash, the others don't.

File: docs/source/debugging.rst (hunk 4 above)
Comment (stuartarchibald) [inline docs/source/debugging.rst]: I agree. I'm also of the view that this manner of operation and explicit ctor should be advertised more generally as the kwargs exposure is useful.

File: docs/source/debugging.rst [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+   >>> from dask.distributed import Client, LocalCluster
+   >>> client = Client()  # This is actually the following two commands
+
+   >>> cluster = LocalCluster()
+   >>> client = Client(cluster.scheduler.address)
+
+LocalCluster is useful because the scheduler and workers are in the same
+process with you, so you can easily inspect their `state
+<http://distributed.readthedocs.io/en/latest/scheduling-state.html>`_ while
+they run (they are running in a separate thread.
```
Comment (stuartarchibald) [inline docs/source/debugging.rst]: Missing trailing bracket.

File: docs/source/debugging.rst [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+
+.. code-block:: python
+
+   >>> cluster.scheduler.processing
+   {'worker-one:59858': {'inc-123', 'add-443'},
+    'worker-two:48248': {'inc-456'}}
+
+You can also do this for the workers *if* you run them without nanny processes
```
Comment (stuartarchibald) [inline docs/source/debugging.rst]: Full stop and end of sentence?

**Comment by stuartarchibald** [comment]
Some general things to perhaps add/discuss:

* Whilst perhaps exceptionally old-school, I often find when all else fails bisection of an error by adding and removing code blocks is sometimes fruitful. Especially when the root cause of the error seems strange.

* Would this be the place for "common mistakes"? Things like forgetting to restart the scheduler/workers having updated the code on disk with e.g. `install`. Writing functions that appear to work locally but cannot be serialized/deserialized across the network? Creating new `delayed` objects from functions running on the workers.

* Should performance debugging go into this document too? If so, perhaps comments about finding out graph sizes, serialization sizes/using cloudpickle to poke at things, what a threaded worker calling a library holding the GIL looks like.

File: docs/source/debugging.rst (hunk 8 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: We currently test this.

```python
@gen_cluster(client=True)
def test_badly_serialized_exceptions(c, s, a, b):
    def f():
        class BadlySerializedException(Exception):
            def __reduce__(self):
                raise TypeError()
        raise BadlySerializedException('hello world')

    x = c.submit(f)

    try:
        result = yield x._result()
    except Exception as e:
        assert 'hello world' in str(e)
    else:
        assert False
```

Can you recommend an alteration to this test that might cover other cases that you've seen?

File: docs/source/debugging.rst (hunk 11 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: I'd like to avoid solutions that require the user to dive into code.

File: docs/source/debugging.rst (hunk 12 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: added link

**Comment by mrocklin** [comment]
Sorry for the delay on this.  I've responded to most of the small comments.  I haven't done anything like add a common mistakes section (I'm not sure I know how to write this without saying a few hundred things) or performance debugging (same).  But hopefully what is here is self contained enough to be a start.

File: docs/source/debugging.rst (hunk 14 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: Fixed

File: docs/source/debugging.rst (hunk 13 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: Fixed

File: docs/source/debugging.rst (hunk 10 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: Fixed

File: docs/source/debugging.rst (hunk 9 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: Fixed

File: docs/source/debugging.rst (hunk 7 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: Fixed

File: docs/source/debugging.rst (hunk 6 above)
Comment (mrocklin) [inline docs/source/debugging.rst]: Fixed

**Comment by mrocklin** [comment]
Merging later today if there are no further comments.

**Pull request merged**